### ISSUE TITLE
Script variables and the Show block bug

### ISSUE THREAD

**Issue by yifatamir** [jmoenig/Snap--Build-Your-Own-Blocks#746]
Today in office hours one of the students I worked with showed me a potential snap bug:

When the "show variable" block is used on a script variable, the variable does not show up on the stage unless the stage is viewed in full-screen mode. 
Also, when the script ends and is run again, the script variable's value that is shown on the stage does not get updated, but rather the "show variable" block must be run again, at which point a new orange bubble containing its value appears on the stage (rather than replacing the old bubble).
![1111](https://cloud.githubusercontent.com/assets/8643357/6704249/599bff9a-cd03-11e4-8a21-38e50d28dc36.JPG)
![222](https://cloud.githubusercontent.com/assets/8643357/6704250/5d144f24-cd03-11e4-80c5-4d4ae27b9a25.JPG)

**Comment by jmoenig** [comment]
Thanks for the report, @yifatamir . I can't reproduce the first behavior you describe. When I create the exact same script and hit the green flag a watcher on the temporary variable shows up, both in full-screen mode and in other stage size modes. Can you tell me which browser and OS you're using (although technically that shouldn't make a difference here)? Also, could you perhaps share a project that shows this issue? Thanks!

Your second issue is actually what I intended. Since the script variable "a" is *temporary* running the script again will create a *new* instance of the variable. Thing is, the *old* temporary variable is actually still around, it just isn't reachable anymore. But - and here's the kicker - you can *closurize* temporary variables and *then* they turn into non-locals and can be indirectly accessed by the closure.

here's a very basic example of of a script variable can "survive" its script and be manipulatable afterwards:

![closure](https://cloud.githubusercontent.com/assets/2596285/6704584/f9d0457e-cd4a-11e4-8d04-9eb609520a08.png)

Does this help?

**Comment by cycomachead** [comment]
I think this report may actually be an issue of when the left column of the stage gets full of variables, the mechanism for telling variables where to show up doesn't always account for the bottom of the stage.

**Comment by jmoenig** [comment]
Oh. Okay, that makes sense, and this also explains why it shows if the stage gets bigger :-)
So, the thing that's useful to know is that executing the "hide variable" block and leaving the variable name slot *empty* removes all watchers on temporary variables, and turn makes way for new ones to be shown *within* the stage.

**Comment by cycomachead** [comment]
> So, the thing that's useful to know is that executing the "hide variable" block and leaving the variable name slot empty removes all watchers on temporary variables, and turn makes way for new ones to be shown within the stage.

I may be misremembering things, but I think I have seen the case where there this doesn't quite fix the issue.

**Comment by brianharvey** [comment]
I vote for "garbage collecting" watchers: when a variable becomes unreachable we should kill the watcher.  I know that's a complication, but I really don't want to give up on two otherwise contradictory goals: (1) the naive user who doesn't know a closure from a banana shouldn't have this sorcerer's-apprentice problem with zombie watchers all over the stage; (2) the advanced user who *does* use closures should be able to debug.

(And the sorcerer's apprentice problem will get worse if I talk you into my compromise idea for making my λ block work, namely that if a second script variable of the same name is created in the same procedure call, instead of just ignoring the second one we make a new variable that shadows the old one.  That way you get the behavior you want (the index of a FOR survives after the loop) and I get what I want (two same-name upvars in the same script not merging).)

**Comment by jmoenig** [comment]
nice idea, Brian, but variables don't know anything about their dependents in Snap, and "garbage collection" at each updateCycle (or in longer intervals) would be too much of an effort and totally drain performance. But we *can* remove all watchers on temporaries whenever the user hits the red stop button...

**Comment by brianharvey** [comment]
But procedures that are about to return know whether or not they're returning a closure, and if not, could kill their temp watchers.

**Comment by jmoenig** [comment]
I don't understand.

Procedures in Snap don't "know" anything about closures until they become evaluated. And even then the *procedure* doesn't know about closures, they're simply somewhere in the environment, not even marked as being temporary. (read the code ^^).

So a "procedure that is about to return" scans its environment for all temporary variables which it *doesn't* return and then "kills their temp watchers". The scanning and killing part takes time and effort, because it happens at *every* procedure call (I can't believe you're seriously suggesting this), and who knows whether a temp var that's unreturned by the current procedure might not be returned by either another case or another procedure?

**Comment by brianharvey** [comment]
No, no, I'm not suggesting it be that perfect.  Just that we preserve all script vars when any closure is returned, but flush them when a non-closure is returned.

**Comment by jmoenig** [comment]
Okay, *that* I do understand, thanks!

So, in my above example the "when green flag pressed" script doesn't "return" anything, so it, too, would kill the temp watcher and thus defeat the purpose of the project, right? Therefore we'd also have to treat any assignment of a value that - somehow (how would we even know?) - originated in a temporary variable frame to a non-temporary construct as a "return". As if that alone weren't hard enough (impossible?), we'd have to let *every* single procedure take a shot at  1) keeping track of its temp watchers and 2) cleaning up obsolete ones. But then, what if two parallel scripts both access a temp var?

![parallel](https://cloud.githubusercontent.com/assets/2596285/6709497/2327ec22-cd7a-11e4-99e8-840033cad95f.png)

wouldn't they "race" to kill it? Oh, and in this particular example would that make the project less debuggable?

I say let's kill temp watchers on (manual!) red-stop-button press! Same as Scratch-like clones.

**Comment by brianharvey** [comment]
You're right about the assignment.  But I still think this is quite workable if you don't try for perfection.  Any time a script creates a closure, it (at that moment) flags its local variables as persistent.  Nothing ever unflags them.  When a script ends, it kills watchers of its unflagged variables.
